### PR TITLE
Add unit tests for getCurrentWeatherUseCase to validate data accuracy and edge case handling

### PR DESCRIPTION
**Pull request by Moha-ALjabouri**

### CODE REVIEW

File: src/main/kotlin/domain/repository/LocationRepository.kt [hunk 1]
```diff
@@ -1,3 +1,7 @@
 package domain.repository
 
-interface LocationRepository
\ No newline at end of file
+import domain.models.Location
+
+interface LocationRepository {
+    suspend fun getLocation(city: String): Location
```
Comment (AimanYosofiUlt) [inline src/main/kotlin/domain/repository/LocationRepository.kt]: The city shouldn't be passed here, the interface should return the location regarding how the implementation is done, please remove the parameter

File: src/main/kotlin/domain/repository/WeatherRepository.kt [hunk 2]
```diff
@@ -1,3 +1,7 @@
 package domain.repository
 
-interface WeatherRepository
\ No newline at end of file
+import domain.models.CurrentWeather
+
+interface WeatherRepository {
+     suspend fun getCurrentWeather(city: String): CurrentWeather
```
Comment (AimanYosofiUlt) [inline src/main/kotlin/domain/repository/WeatherRepository.kt]: same here, please remove the ciye parameter

File: src/main/kotlin/data/repository/WeatherRepositoryImpl.kt [hunk 3]
```diff
@@ -1,5 +1,11 @@
 package data.repository
 
+import domain.models.CurrentWeather
 import domain.repository.WeatherRepository
 
-class WeatherRepositoryImpl : WeatherRepository
\ No newline at end of file
+class WeatherRepositoryImpl : WeatherRepository{
```
Comment (AimanYosofiUlt) [inline src/main/kotlin/data/repository/WeatherRepositoryImpl.kt]: Here you should be passed an HTTPClient that give you the weather, and location too
`WeatherRepositoryImpl(client: HTTPClient, location:Location)`

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    )
+
+    @BeforeEach
+    fun setup() {
+        weatherRepository = mockk(relaxed = true)
+        getCurrentWeatherUseCase = GetCurrentWeatherUseCase(weatherRepository)
+    }
+
+    @Test
+    fun `should return current weather for given city`() = runTest {
```
Comment (AimanYosofiUlt) [inline src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Please keep the convention of **should when** please
You can convert it to "should return current weather for given city when getting the current weather"

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        // Given
+        coEvery { weatherRepository.getCurrentWeather(city) } returns expectedWeather
+
+        // When
+        val result = getCurrentWeatherUseCase(city)
+
+        // Then
+        assertEquals(expectedWeather, result)
```
Comment (AimanYosofiUlt) [inline src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Keep it with one assertion, it's good practice to follow
Notice that: assertEquals is an assertion, but coVerify is an assertion too (**tow assertions here**)

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Test
+    fun `should throw exception when repository fails`() = runTest {
+        // Given
+        val exceptionMessage = "API error"
+        coEvery { weatherRepository.getCurrentWeather(city) } throws RuntimeException(exceptionMessage)
+
+        // When / Then
+        try {
+            getCurrentWeatherUseCase(city)
```
Comment (AimanYosofiUlt) [inline src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Please don't use try-catch, use assertThrow instead

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    fun `should return weather with valid temperature value`() = runTest {
+        // Given
+        val modifiedWeather = expectedWeather.copy(temperature2m = "17.8")
+        coEvery { weatherRepository.getCurrentWeather(city) } returns modifiedWeather
+
+        // When
+        val result = getCurrentWeatherUseCase(city)
+
+        // Then
+        assertEquals("17.8", result.temperature2m)
```
Comment (AimanYosofiUlt) [inline src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: For consistency, please the Truth library, we will use it in the whole project

**Review by AimanYosofiUlt — COMMENTED** [review]
Please see the comments, and give me a feedback about them or resolve the issues
Good job though

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    )
+
+    @BeforeEach
+    fun setup() {
+        weatherRepository = mockk(relaxed = true)
+        getCurrentWeatherUseCase = GetCurrentWeatherUseCase(weatherRepository)
+    }
+
+    @Test
+    fun `should return current weather for given coordinates when getting the current weather`() = runTest {
```
Comment (aabukmail) [inline src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt:54]: Add getCurrentWeather() in the beginning of fun + edit it to be like this: "getCurrentWeather() should return current weather when getting the location"

File: src/main/kotlin/domain/repository/LocationRepository.kt (hunk 1 above)
Comment (Moha-ALjabouri) [reply src/main/kotlin/domain/repository/LocationRepository.kt]: Done

File: src/main/kotlin/domain/repository/WeatherRepository.kt (hunk 2 above)
Comment (Moha-ALjabouri) [reply src/main/kotlin/domain/repository/WeatherRepository.kt]: Done

File: src/main/kotlin/data/repository/WeatherRepositoryImpl.kt (hunk 3 above)
Comment (Moha-ALjabouri) [reply src/main/kotlin/data/repository/WeatherRepositoryImpl.kt]: Done

**Review by aabukmail — COMMENTED** [review]
Add getCurrentWeather() in the beginning of fun + edit it to be like this: "getCurrentWeather() should return current weather when getting the location"

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt (hunk 8 above)
Comment (Moha-ALjabouri) [reply src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt:54]: Done

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt (hunk 4 above)
Comment (Moha-ALjabouri) [reply src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Done

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt (hunk 5 above)
Comment (Moha-ALjabouri) [reply src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Done

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt (hunk 6 above)
Comment (Moha-ALjabouri) [reply src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Done

File: src/main/kotlin/data/repository/LocationRepositoryImpl.kt [hunk 9]
```diff
@@ -1,5 +1,14 @@
 package data.repository
 
+import domain.models.Location
 import domain.repository.LocationRepository
 
-class LocationRepositoryImpl : LocationRepository
\ No newline at end of file
+class LocationRepositoryImpl : LocationRepository {
+    override suspend fun getCurrentLocation(): Location {
+        return Location(
```
Comment (AimanYosofiUlt) [inline src/main/kotlin/data/repository/LocationRepositoryImpl.kt]: Why return values here?, and the location shouldn't be your responsibility 😁

File: src/main/kotlin/data/repository/WeatherRepositoryImpl.kt [hunk 10]
```diff
@@ -1,5 +1,29 @@
...
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
 
-class WeatherRepositoryImpl : WeatherRepository
\ No newline at end of file
+class WeatherRepositoryImpl(private val httpClient: HttpClient) : WeatherRepository {
+    override suspend fun getCurrentWeather(latitude: Double, longitude: Double): CurrentWeather {
+        val response: HttpResponse =
+            httpClient.get("https://my-server.tld/v1/forecast?latitude=52.52&longitude=13.41&hourly=temperature_2m&current=temperature_2m,relative_humidity_2m,apparent_temperature,is_day,precipitation,showers,weather_code,cloud_cover,wind_speed_10m") {
```
Comment (AimanYosofiUlt) [inline src/main/kotlin/data/repository/WeatherRepositoryImpl.kt]: store it in a constant variable please>>>

File: src/main/kotlin/data/repository/WeatherRepositoryImpl.kt [hunk 11]
```diff
@@ -1,5 +1,29 @@
...
+import domain.models.CurrentWeather
+import io.ktor.client.*
+import io.ktor.client.call.*
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
 
-class WeatherRepositoryImpl : WeatherRepository
\ No newline at end of file
+class WeatherRepositoryImpl(private val httpClient: HttpClient) : WeatherRepository {
```
Comment (AimanYosofiUlt) [inline src/main/kotlin/data/repository/WeatherRepositoryImpl.kt:14]: store it in a constant, like this
`const val OPEN_METO_API = https://my-server.tld/v1/forecast`

File: src/main/kotlin/data/repository/WeatherRepositoryImpl.kt [hunk 12]
```diff
@@ -1,5 +1,29 @@
...
+                parameter("latitude", latitude)
+                parameter("longitude", longitude)
+                parameter(
+                    "current", "temperature_2m,relative_humidity_2m,apparent_temperature,is_day," +
+                            "precipitation,showers,weather_code,cloud_cover,wind_speed_10m"
+                )
+            }
+
+        if (!response.status.isSuccess()) {
+            throw Exception("Failed to fetch weather data: ${response.status}")
```
Comment (AimanYosofiUlt) [inline src/main/kotlin/data/repository/WeatherRepositoryImpl.kt]: Use a custom exception instead of a generic one

File: src/main/kotlin/domain/use_cases/GetCurrentWeatherUseCase.kt [hunk 13]
```diff
@@ -1,9 +1,6 @@
...
-
-class GetCurrentWeatherUseCase(
-    private val weatherRepository: WeatherRepository
-) {
-
-}
\ No newline at end of file
+import domain.models.CurrentWeather
+class GetCurrentWeatherUseCase(private val repository: WeatherRepository) {
+    suspend operator fun invoke(latitude: Double, longitude: Double): CurrentWeather {
```
Comment (AimanYosofiUlt) [inline src/main/kotlin/domain/use_cases/GetCurrentWeatherUseCase.kt]: Rename it to a more proper name, like getCurrentWeather()

File: src/main/kotlin/domain/use_cases/GetLocationUseCase.kt [hunk 14]
```diff
@@ -1,6 +1,10 @@
 package domain.use_cases
 
-import domain.util.location_getter.LocationFetcher
+import domain.models.Location
+import domain.repository.LocationRepository
 
-class GetLocationUseCase(private val locationGetter: LocationFetcher) {
+class GetLocationUseCase(private val repository: LocationRepository) {
+    suspend fun execute(): Location {
```
Comment (AimanYosofiUlt) [inline src/main/kotlin/domain/use_cases/GetLocationUseCase.kt]: Rename it to a more proper name, like getLocation()

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Test
+    fun `should throw exception when repository fails`() = runTest {
+        // Given
+        val exceptionMessage = "API error"
+        coEvery { weatherRepository.getCurrentWeather(latitude, longitude) } throws RuntimeException(exceptionMessage)
+
+        try {
+            // When
+            getCurrentWeatherUseCase(latitude, longitude)
```
Comment (AimanYosofiUlt) [inline src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Please don't use try-catch here to assert the exception, you can use assertThrow{} instead

File: src/main/kotlin/data/repository/WeatherRepositoryImpl.kt [hunk 16]
```diff
@@ -1,5 +1,25 @@
...
+class WeatherRepositoryImpl(private val httpClient: HttpClient) : WeatherRepository {
+    override suspend fun getCurrentWeather(latitude: Double, longitude: Double): CurrentWeather {
+        val response: HttpResponse = httpClient.get("https://api.open-meteo.com/v1/forecast") {
+            parameter("latitude", latitude)
+            parameter("longitude", longitude)
+            parameter("current", "temperature_2m,relative_humidity_2m,apparent_temperature,is_day,wind_speed_10m,snowfall,rain,weather_code,cloud_cover")
+        }
+
+        if (!response.status.isSuccess()) {
+            throw Exception("Failed to fetch weather data: ${response.status}")
```
Comment (Hasa9nAli) [inline src/main/kotlin/data/repository/WeatherRepositoryImpl.kt]: i think use custom exception is better way as -> `FailedFetchWeatherDataException`

**Review by Hasa9nAli — COMMENTED** [review]
Good work keep going @Moha-ALjabouri

**Review by AimanYosofiUlt — COMMENTED** [review]
good job bro

File: src/main/kotlin/data/repository/WeatherRepositoryImpl.kt (hunk 12 above)
Comment (Moha-ALjabouri) [reply src/main/kotlin/data/repository/WeatherRepositoryImpl.kt]: Done

File: src/main/kotlin/data/repository/WeatherRepositoryImpl.kt (hunk 11 above)
Comment (Moha-ALjabouri) [reply src/main/kotlin/data/repository/WeatherRepositoryImpl.kt:14]: done

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt (hunk 15 above)
Comment (Moha-ALjabouri) [reply src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: done

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        // When
+        val result = getCurrentWeatherUseCase(latitude, longitude)
+
+        // Then
+        coVerify(exactly = 1) { weatherRepository.getCurrentWeather(latitude, longitude) }
+        assertEquals(expectedWeather, result)
+    }
+
+    @Test
+    fun `should throw exception when repository fails`() = runTest {
```
Comment (aabukmail) [inline src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Please add a call to getCurrentWeather() at the beginning of the function. Also, use the following syntax for your test descriptions:
your_fun() should return ... when ...

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+            getCurrentWeatherUseCase(latitude, longitude)
+            fail("Exception was expected")
+        } catch (e: RuntimeException) {
+            // Then
+            assertThat(e.message).isEqualTo(exceptionMessage)
+        }
+    }
+
+    @Test
+    fun `should return weather with valid temperature value`() = runTest {
```
Comment (aabukmail) [inline src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Please add a call to getCurrentWeather() at the beginning of the function. Also, use the following syntax for your test descriptions:
your_fun() should return ... when ...

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+        // When
+        val result = getCurrentWeatherUseCase(latitude, longitude)
+
+        // Then
+        assertThat(result.temperature2m).isEqualTo(20.8)
+    }
+
+    @Test
+    fun `should call repository with zero coordinates`() = runTest {
```
Comment (aabukmail) [inline src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Please add a call to getCurrentWeather() at the beginning of the function. Also, use the following syntax for your test descriptions:
your_fun() should return ... when ...

File: src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt (hunk 18 above)
Comment (aabukmail) [reply src/test/kotlin/domain/use_cases/GetCurrentWeatherUseCaseTest.kt]: Why did you choose to use only temperature as a key value for this test, without considering other parameters?
I believe your case should handle latitude and longitude to return the corresponding weather data.

**Review by aabukmail — CHANGES_REQUESTED** [review]

**Review by AimanYosofiUlt — APPROVED** [review]
Good work guys
